Store now keeps a weak reference to subscribers

vfn opened the pull request:
Transformed the typealias `Subscription` into a `Struct` and added a weak reference to the `subscriber`.

Fixes https://github.com/ReSwift/ReSwift/issues/62

agentk commented:
Hi @vfn,
Changes look good. Exactly how I was planning on ensuring weak subscriber refs.
Before we merge the commit could you add a test for ensuring dereferenced subscribers are removed from the subscriptions array. This will also keep coverage complete.
Once complete just squash the commits to keep the history concise.
Cheers, Karl. P.s. nice to see someone so close (by Aussie distances at least :-)

Ben-G commented on ReSwift/CoreTypes/Subscription.swift:
```diff
@@ -0,0 +1,14 @@
+struct Subscription<State: StateType> {
+    private(set) weak var subscriber: AnyStoreSubscriber? = nil
+    private(set) var selector: (State -> Any)?
+}
```
Can we just make these `let` instead of `private(set)`?

Ben-G commented on ReSwift/CoreTypes/Store.swift:
```diff
@@ -17,14 +17,26 @@ import Foundation
+            for (index, subscription) in subscriptions.enumerate() {
+                guard let subscriber = subscription.subscriber else {
+                    indexesToRemove.append(index)
+                    return
```
The escape route here should be a `continue` and not a `return`. With the current code subscribers after a no-longer-existent subscriber would miss a state update.

I think it would be best to separate these two pieces of logic:
1) Filter out subscriptions that have lost their reference to a subscriber
2) Notify all remaining subscribers

Could look something like this:

```
subscriptions = subscriptions.filter { $0.subscriber != nil }
subscriptions.forEach {
     $0.subscriber!._newState(subscription.selector?(state) ?? state)
}
```
Since the use of the implicit unwrapped optional here is very close to the line that filters out nil-values, I think it is fine to use and shouldn't through off folks reading the code.

vfn replied:
@Ben-G, I like how the code is concise with this approach, although I'm not sure if the best approach here is to run the filter on every attempt to notify the subscribers.

Ben-G replied:
If you like we can look into some performance test, but the operation
should be extremely cheap for all use cases I can think of. I'm also open
to other suggestions :) Ideally though I would like to treat these to tasks
separately to make the code easier to understand and to eliminate the risk
of subtle changes that could cause bugs.

--
Benjamin Encz

Find me on twitter: @benjaminencz <https://twitter.com/benjaminencz>

On Thu, Feb 4, 2016 at 1:46 PM, Virgilio Favero Neto <
notifications@github.com> wrote:

> In ReSwift/CoreTypes/Store.swift
> <https://github.com/ReSwift/ReSwift/pull/70#discussion_r51942915>:
>
> >      // TODO: Setter should not be public; need way for store enhancers to modify appState anyway
> >
> >      /*private (set)*/ public var state: State! {
> >          didSet {
> > -            subscriptions.forEach {
> > +            var indexesToRemove: [Int] = []
> > +            for (index, subscription) in subscriptions.enumerate() {
> > +                guard let subscriber = subscription.subscriber else {
> > +                    indexesToRemove.append(index)
> > +                    return
>
> @Ben-G <https://github.com/Ben-G>, I like how the code is concise with
> this approach, although I'm not sure if the best approach here is to run
> the filter on every attempt to notify the subscribers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ReSwift/ReSwift/pull/70/files#r51942915>.
>

vfn replied:
I'm happy with your suggestion.

I'm leaning towards keeping the optional chaining when accessing the subscriber instead of implicit unwrapping it. There is no cost for the call, and it saves potential crashes caused by subscribers being deinitalized while in the `forEach`. Does that make sense?

Ben-G replied:
Ah yes, that makes more sense. There's indeed no need to implicitly unwrap
here! Optional chaining it is!

--
Benjamin Encz

Find me on twitter: @benjaminencz <https://twitter.com/benjaminencz>

On Thu, Feb 4, 2016 at 2:42 PM, Virgilio Favero Neto <
notifications@github.com> wrote:

> In ReSwift/CoreTypes/Store.swift
> <https://github.com/ReSwift/ReSwift/pull/70#discussion_r51950264>:
>
> >      // TODO: Setter should not be public; need way for store enhancers to modify appState anyway
> >
> >      /*private (set)*/ public var state: State! {
> >          didSet {
> > -            subscriptions.forEach {
> > +            var indexesToRemove: [Int] = []
> > +            for (index, subscription) in subscriptions.enumerate() {
> > +                guard let subscriber = subscription.subscriber else {
> > +                    indexesToRemove.append(index)
> > +                    return
>
> I'm happy with your suggestion.
>
> I'm leaning towards keeping the optional chaining when accessing the
> subscriber instead of implicit unwrapping it. There is no cost for the
> call, and it saves potential crashes caused by subscribers being
> deinitalized while in the forEach. Does that make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ReSwift/ReSwift/pull/70/files#r51950264>.
>

Ben-G commented:
Hey @vfn, great contribution. Thanks a lot!

See my comments for some issues and some suggested changes. When this is ready to merge we should have test cases for:

1. A subscriber that is no longer referenced strongly should become deallocated, even when subscribed to the store
2. Other subscriptions still work expected, even when a subscription is cleaned up during the next notification cycle

It also seems like this here would conflict with #71. The change in #71 is great, but it probably makes sense to combine the changes into this PR, what do you think?

Lastly, if you don't have time/interest to make the necessary changes, you can also re-open a PR against `dev` and we can take the rest of the implementation over from there. Whatever you prefer :+1: 

Thanks a lot for the contribution :fireworks:

vfn commented:
Hi @Ben-G and @agentk, thank you for reviewing the PR so quickly. I'm working on the fixes and I'll update the PR in the next few hours.

Do you want me to push the fixes as fixup commits and once it all looks good I/we squash them, before merging, or do you prefer to have the PR already updated with the changes squashed?

Ben-G commented:
I think it makes sense to review first and squash afterwards! That way we
can save you some time :)

Thanks a lot!

--
Benjamin Encz

Find me on twitter: @benjaminencz <https://twitter.com/benjaminencz>

On Thu, Feb 4, 2016 at 1:43 PM, Virgilio Favero Neto <
notifications@github.com> wrote:

> Hi @Ben-G <https://github.com/Ben-G> and @agentk
> <https://github.com/agentk>, thank you for reviewing the PR so quickly.
> I'm working on the fixes and I'll update the PR in the next few hours.
>
> Do you want me to push the fixes as fixup commits and once it all looks
> good I/we squash them, before merging, or do you prefer to have the PR
> already updated with the changes squashed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ReSwift/ReSwift/pull/70#issuecomment-180064529>.
>

vfn commented:
@Ben-G @garnett I have updated the PR with the suggested changes, added the tests and also added the commit from PR https://github.com/ReSwift/ReSwift/pull/71 into this PR.

Unfortunatelly I screwed up and created this PR against `master`, instead of `dev`. Once we finish reviewing the stuff here, I can create another PR agains `dev` and we can reference the discussion on this PR here that can be closed afterwards.

Ben-G commented:
@vfn This looks great now! Thanks. You can go ahead and rebase now :smiley: 

In this state it's fine to merge directly into `master`. We can always add a test case or two later, but IMO coverage is sufficient for now!

vfn commented:
@Ben-G I've rebase and squashed the fixups. Let me know if you need anything to be done  before merging it

Ben-G commented:
@vfn This looks great! Thanks a lot for the contribution :star2: :star2: 

As a contributor I'll invite you to the organization. Feel free to ignore the invite if you're not interested.

Cheers!
